README.md
# Rinvex Language

**Rinvex Language** is a simple and lightweight package for retrieving language details with flexibility. A whole bunch of data including name, native, iso codes, language family, language script, language cultures, and other attributes for the 180+ known languages worldwide at your fingertips.

[![Packagist](https://img.shields.io/packagist/v/rinvex/languages.svg?label=Packagist&style=flat-square)](https://packagist.org/packages/rinvex/languages)
[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/rinvex/languages.svg?label=Scrutinizer&style=flat-square)](https://scrutinizer-ci.com/g/rinvex/languages/)
[![Travis](https://img.shields.io/travis/rinvex/languages.svg?label=TravisCI&style=flat-square)](https://travis-ci.org/rinvex/languages)
[![StyleCI](https://styleci.io/repos/77772990/shield)](https://styleci.io/repos/77772990)
[![License](https://img.shields.io/packagist/l/rinvex/languages.svg?label=License&style=flat-square)](https://github.com/rinvex/languages/blob/develop/LICENSE)


## Usage

Install via `composer require rinvex/languages`, then use intuitively:

```php
// Get single language
$english = language('en');

// Get language name: English
echo $english->getName();

// Get language native name: English
echo $english->getNativeName();

// Get language ISO 639-1 code: en
echo $english->getIso6391();

// Get language ISO 639-2 code: eng
echo $english->getIso6392();

// Get language ISO 639-3 code: eng
echo $english->getIso6393();

// Get language script details: {"name": "Latin","iso_15924": "Latn","iso_numeric": "215","direction": "ltr"}
echo $english->getScript();

// Get language script name: Latin
echo $english->getScriptName();

// Get language script ISO 15924 code: Latn
echo $english->getScriptIso15924();

// Get language script ISO numeric code: 215
echo $english->getScriptIsoNumeric();

// Get language script direction: ltr
echo $english->getScriptDirection();

// Get language family details: {"name": "Indo-European","iso_639_5": "ine","hierarchy": "ine"}
echo $english->getFamily();

// Get language family name: Indo-European
echo $english->getFamilyName();

// Get language family ISO 6395 code: ine
echo $english->getFamilyIso6395();

// Get language family hierarchy: ine
echo $english->getFamilyHierarchy();

// Get language scope: individual
echo $english->getScope();

// Get language type: living
echo $english->getType();

// Get language cultures: {"en-US": {"name": "English (United States)","native": "English (United States)"}, {...}}
echo $english->getCultures();

// Get language specific culture: {"name": "English (United States)","native": "English (United States)"}
echo $english->getCulture('en-US');


// Get all languages
$languages = languages();

// Get all language scripts
$language_scripts = language_scripts();

// Get all language families
$language_families = language_families();

// Get languages with where condition (language script: Latin)
$whereLanguages = \Rinvex\Language\LanguageLoader::where('script.name', 'Latin');
```

> **Notes:**
> - **Rinvex Language** is framework-agnostic, so it's compatible with any PHP framework whatsoever without any dependencies at all, except for the PHP version itself **^8.0.0**. Awesome, huh? :smiley:
> - **Rinvex Language** provides the global helpers for your convenience and for ease of use, but in fact it's just wrappers around the underlying `LanguageLoader` class, which you can utilize and use directly if you wish


## Features Explained

- Language data are all stored here: `resources/languages.json`.
- `name` - language english name
- `native` - language native name
- `iso_639_1` - two letter code ISO 639-1
- `iso_639_2` - three letter code ISO 639-2
- `iso_639_3` - three letter code ISO 639-3
- `script` - language script details
    - name: language script name
    - iso_15924: language script ISO 15924 code
    - iso_numeric: language script ISO numeric code
    - direction: language script writing direction
- `family` - language family details
    - name: language family name
    - iso_639_5: three-letter ISO 639-5 code
    - hierarchy: language family hierarchy
- `cultures` - list of language cultures
    - key: four-letter language culture code (iso_639_1-iso_3166_1)
    - value: culture object
        - name: language culture name
        - native: language culture native name
- `scope` - language scope (like individual or macrolanguage)
- `type` - language type (like living or ancient)


## Changelog

Refer to the [Changelog](CHANGELOG.md) for a full history of the project.


## Support

The following support channels are available at your fingertips:

- [Chat on Slack](https://bit.ly/rinvex-slack)
- [Help on Email](mailto:help@rinvex.com)
- [Follow on Twitter](https://twitter.com/rinvex)


## Contributing & Protocols

Thank you for considering contributing to this project! The contribution guide can be found in [CONTRIBUTING.md](CONTRIBUTING.md).

Bug reports, feature requests, and pull requests are very welcome.

- [Versioning](CONTRIBUTING.md#versioning)
- [Pull Requests](CONTRIBUTING.md#pull-requests)
- [Coding Standards](CONTRIBUTING.md#coding-standards)
- [Feature Requests](CONTRIBUTING.md#feature-requests)
- [Git Flow](CONTRIBUTING.md#git-flow)


## Security Vulnerabilities

If you discover a security vulnerability within this project, please send an e-mail to [help@rinvex.com](help@rinvex.com). All security vulnerabilities will be promptly addressed.


## About Rinvex

Rinvex is a software solutions startup, specialized in integrated enterprise solutions for SMEs established in Alexandria, Egypt since June 2016. We believe that our drive The Value, The Reach, and The Impact is what differentiates us and unleash the endless possibilities of our philosophy through the power of software. We like to call it Innovation At The Speed Of Life. That’s how we do our share of advancing humanity.


## License

This software is released under [The MIT License (MIT)](LICENSE).

(c) 2016-2022 Rinvex LLC, Some rights reserved.


CHANGELOG.md
# Rinvex Languages Change Log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](CONTRIBUTING.md).


## [v7.0.1] - 2023-05-02
- 3c95156: Update phpunit to v10.1 from v9.5

## [v7.0.0] - 2023-01-09
- Drop PHP v8.0 support and update composer dependencies

## [v6.0.1] - 2022-05-01
- Fix language validator rule to check if the key is 2 char string

## [v6.0.0] - 2021-08-22
- Drop PHP v7 support
- Update composer dependencies
- Enable StyleCI risky mode

## [v5.0.3] - 2020-12-25
- Add support for PHP v8

## [v5.0.2] - 2020-12-22
- Update composer dependencies
- Drop PHP 7.2 & 7.3 support from travis
- Replace hardcoded phase with language phrase
- Remove default indent size config

## [v5.0.1] - 2020-04-04
- Drop laravel/helpers usage as it's no longer used

## [v5.0.0] - 2020-03-15
- Upgrade to Laravel v7.1.x & PHP v7.4.x

## [v4.1.2] - 2020-03-13
- Tweak TravisCI config
- Add migrations autoload option to the package
- Tweak service provider `publishesResources`
- Update StyleCI config

## [v4.1.1] - 2019-09-24
- Add missing laravel/helpers composer package

## [v4.1.0] - 2019-06-02
- Update composer deps
- Drop PHP 7.1 travis test

## [v4.0.0] - 2019-03-03
- Rename environment variable QUEUE_DRIVER to QUEUE_CONNECTION
- Require PHP 7.2 & Laravel 5.8
- Apply PHPUnit 8 updates

## [v3.0.1] - 2018-12-22
- Update composer dependencies
- Add PHP 7.3 support to travis

## [v3.0.0] - 2018-10-01
- Enforce Consistency
- Rename package to rinvex/languages

## [v2.0.1] - 2018-09-22
- Update travis php versions
- Add Laravel Auto Discovery support and validation rule
- Drop StyleCI multi-language support (paid feature now!)
- Update composer dependencies
- Prepare and tweak testing configuration
- Fix wrong readme php version requirement (fix #7)
- Update StyleCI options
- Update PHPUnit options

## [v2.0.0] - 2018-02-18
- Require PHP v7.1.3
- Update supplementary files
- Update composer dependencies
- Fix deprecated PHPUnit TestCase namespace
- Add PHPUnitPrettyResultPrinter
- Typehint method returns

## [v1.1.0] - 2017-03-07
- Exclude resources directory from CI
- Rename helper test suites
- Update StyleCI fixers and other supplementary files
- Enforce strict comparison
- Enforce strict type declaration
- Convert language code into small case before processing
- Change internal method visibility to protected
- Execute reflections once per test class

## v1.0.0 - 2017-01-01
- Tag first release

[v7.0.1]: https://github.com/rinvex/languages/compare/v7.0.0...v7.0.1
[v7.0.0]: https://github.com/rinvex/languages/compare/v6.0.1...v7.0.0
[v6.0.1]: https://github.com/rinvex/languages/compare/v6.0.0...v6.0.1
[v6.0.0]: https://github.com/rinvex/languages/compare/v5.0.3...v6.0.0
[v5.0.3]: https://github.com/rinvex/languages/compare/v5.0.2...v5.0.3
[v5.0.2]: https://github.com/rinvex/languages/compare/v5.0.1...v5.0.2
[v5.0.1]: https://github.com/rinvex/languages/compare/v5.0.0...v5.0.1
[v5.0.0]: https://github.com/rinvex/languages/compare/v4.1.2...v5.0.0
[v4.1.2]: https://github.com/rinvex/languages/compare/v4.1.1...v4.1.2
[v4.1.1]: https://github.com/rinvex/languages/compare/v4.1.0...v4.1.1
[v4.1.0]: https://github.com/rinvex/languages/compare/v4.0.0...v4.1.0
[v4.0.0]: https://github.com/rinvex/languages/compare/v3.0.1...v4.0.0
[v3.0.1]: https://github.com/rinvex/languages/compare/v3.0.0...v3.0.1
[v3.0.0]: https://github.com/rinvex/languages/compare/v2.0.1...v3.0.0
[v2.0.1]: https://github.com/rinvex/languages/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/rinvex/languages/compare/v1.1.0...v2.0.0
[v1.1.0]: https://github.com/rinvex/languages/compare/v1.0.0...v1.1.0


tests/HelpersTest.php
<?php

declare(strict_types=1);

namespace Rinvex\Language\Tests;

use Rinvex\Language\Language;
use PHPUnit\Framework\TestCase;

class HelpersTest extends TestCase
{
    /** @test */
    public function it_returns_language_data_through_helper()
    {
        $amharic = [
            'name' => 'Amharic',
            'native' => 'አማርኛ',
            'iso_639_1' => 'am',
            'iso_639_2' => 'amh',
            'iso_639_3' => 'amh',
            'script' => [
                'name' => 'Ethiopic (Ge_ez)',
                'iso_15924' => 'Ethi',
                'iso_numeric' => '430',
                'direction' => 'ltr',
            ],
            'family' => [
                'name' => 'Afro-Asiatic',
                'iso_639_5' => 'afa',
                'hierarchy' => 'afa',
            ],
            'cultures' => [
                'am-ET' => [
                    'name' => 'Amharic (Ethiopia)',
                    'native' => 'አማርኛ (ኢትዮጵያ)',
                ],
            ],
            'scope' => 'individual',
            'type' => 'living',
        ];

        $this->assertEquals($amharic, language('am', false));
        $this->assertEquals(new Language($amharic), language('am'));
    }

    /** @test */
    public function it_returns_languages_array_through_helper()
    {
        $this->assertEquals(183, count(languages()));
        $this->assertInternalType('array', languages()['en']);
        $this->assertEquals('English', languages()['en']['name']);
    }

    /** @test */
    public function it_returns_language_scripts_array_through_helper()
    {
        $this->assertEquals(29, count(language_scripts()));
        $this->assertInternalType('array', language_scripts());
        $this->assertArrayHasKey('Arab', language_scripts());
    }

    /** @test */
    public function it_returns_language_families_array_through_helper()
    {
        $this->assertEquals(27, count(language_families()));
        $this->assertInternalType('array', language_families());
        $this->assertArrayHasKey('afa', language_families());
    }
}


tests/LanguageTest.php
<?php

declare(strict_types=1);

namespace Rinvex\Language\Tests;

use Exception;
use Rinvex\Language\Language;
use PHPUnit\Framework\TestCase;

class LanguageTest extends TestCase
{
    /** @var array */
    protected $languageArray;

    /** @var \Rinvex\Language\Language */
    protected $languageObject;

    public function setUp()
    {
        parent::setUp();

        $this->languageArray = [
            'name' => 'Amharic',
            'native' => 'አማርኛ',
            'iso_639_1' => 'am',
            'iso_639_2' => 'amh',
            'iso_639_3' => 'amh',
            'script' => [
                'name' => 'Ethiopic (Ge_ez)',
                'iso_15924' => 'Ethi',
                'iso_numeric' => '430',
                'direction' => 'ltr',
            ],
            'family' => [
                'name' => 'Afro-Asiatic',
                'iso_639_5' => 'afa',
                'hierarchy' => 'afa',
            ],
            'cultures' => [
                'am-ET' => [
                    'name' => 'Amharic (Ethiopia)',
                    'native' => 'አማርኛ (ኢትዮጵያ)',
                ],
            ],
            'scope' => 'individual',
            'type' => 'living',
        ];

        $this->languageObject = new Language($this->languageArray);
    }

    /** @test */
    public function it_throws_an_exception_when_missing_mandatory_attributes()
    {
        $this->expectException(Exception::class);

        new Language([]);
    }

    /** @test */
    public function it_sets_attributes_once_instantiated()
    {
        $this->assertEquals($this->languageArray['name'], $this->languageObject->getName());
        $this->assertEquals($this->languageArray['native'], $this->languageObject->getNativeName());
        $this->assertEquals($this->languageArray['iso_639_1'], $this->languageObject->getIso6391());
    }

    /** @test */
    public function it_gets_attributes()
    {
        $this->assertEquals($this->languageArray, $this->languageObject->getAttributes());
    }

    /** @test */
    public function it_sets_attributes()
    {
        $this->languageObject->setAttributes(['iso_639_3' => 'cha']);

        $this->assertEquals('cha', $this->languageObject->getIso6393());
    }

    /** @test */
    public function it_gets_dotted_attribute()
    {
        $this->assertEquals($this->languageArray['script']['iso_15924'], $this->languageObject->get('script.iso_15924'));
    }

    /** @test */
    public function it_gets_default_when_missing_value()
    {
        $this->assertEquals('default', $this->languageObject->get('unknown', 'default'));
    }

    /** @test */
    public function it_gets_all_attributes_when_missing_key()
    {
        $this->assertEquals($this->languageArray, $this->languageObject->get(null));
    }

    /** @test */
    public function it_sets_attribute()
    {
        $this->languageObject->set('iso_639_3', 'cha');

        $this->assertEquals('cha', $this->languageObject->getIso6393());
    }

    /** @test */
    public function its_fluently_chainable_when_sets_attributes()
    {
        $this->assertEquals($this->languageObject, $this->languageObject->setAttributes([]));
    }

    /** @test */
    public function it_returns_name()
    {
        $this->assertEquals($this->languageArray['name'], $this->languageObject->getName());
    }

    /** @test */
    public function it_returns_null_when_missing_name()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getName());
    }

    /** @test */
    public function it_returns_native_name()
    {
        $this->assertEquals($this->languageArray['native'], $this->languageObject->getNativeName());
    }

    /** @test */
    public function it_returns_null_when_missing_native_name()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getNativeName());
    }

    /** @test */
    public function it_returns_iso6391()
    {
        $this->assertEquals($this->languageArray['iso_639_1'], $this->languageObject->getIso6391());
    }

    /** @test */
    public function it_returns_null_when_missing_iso6391()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getIso6391());
    }

    /** @test */
    public function it_returns_iso6392()
    {
        $this->assertEquals($this->languageArray['iso_639_2'], $this->languageObject->getIso6392());
    }

    /** @test */
    public function it_returns_null_when_missing_iso6392()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getIso6392());
    }

    /** @test */
    public function it_returns_iso6393()
    {
        $this->assertEquals($this->languageArray['iso_639_3'], $this->languageObject->getIso6393());
    }

    /** @test */
    public function it_returns_null_when_missing_iso6393()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getIso6393());
    }

    /** @test */
    public function it_returns_script()
    {
        $this->assertEquals($this->languageArray['script'], $this->languageObject->getScript());
    }

    /** @test */
    public function it_returns_null_when_missing_script()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScript());
    }

    /** @test */
    public function it_returns_script_name()
    {
        $this->assertEquals($this->languageArray['script']['name'], $this->languageObject->getScriptName());
    }

    /** @test */
    public function it_returns_null_when_missing_script_name()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScriptName());
    }

    /** @test */
    public function it_returns_script_iso_15924()
    {
        $this->assertEquals($this->languageArray['script']['iso_15924'], $this->languageObject->getScriptIso15924());
    }

    /** @test */
    public function it_returns_null_when_missing_script_iso_15924()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScriptIso15924());
    }

    /** @test */
    public function it_returns_script_iso_numeric()
    {
        $this->assertEquals($this->languageArray['script']['iso_numeric'], $this->languageObject->getScriptIsoNumeric());
    }

    /** @test */
    public function it_returns_null_when_missing_script_iso_numeric()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScriptIsoNumeric());
    }

    /** @test */
    public function it_returns_script_direction()
    {
        $this->assertEquals($this->languageArray['script']['direction'], $this->languageObject->getScriptDirection());
    }

    /** @test */
    public function it_returns_null_when_missing_script_direction()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScriptDirection());
    }

    /** @test */
    public function it_returns_family()
    {
        $this->assertEquals($this->languageArray['family'], $this->languageObject->getFamily());
    }

    /** @test */
    public function it_returns_null_when_missing_family()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getFamily());
    }

    /** @test */
    public function it_returns_family_name()
    {
        $this->assertEquals($this->languageArray['family']['name'], $this->languageObject->getFamilyName());
    }

    /** @test */
    public function it_returns_null_when_missing_family_name()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getFamilyName());
    }

    /** @test */
    public function it_returns_family_iso_639_5()
    {
        $this->assertEquals($this->languageArray['family']['iso_639_5'], $this->languageObject->getFamilyIso6395());
    }

    /** @test */
    public function it_returns_null_when_missing_family_iso_639_5()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getFamilyIso6395());
    }

    /** @test */
    public function it_returns_family_hierarchy()
    {
        $this->assertEquals($this->languageArray['family']['hierarchy'], $this->languageObject->getFamilyHierarchy());
    }

    /** @test */
    public function it_returns_null_when_missing_family_hierarchy()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getFamilyHierarchy());
    }

    /** @test */
    public function it_returns_scope()
    {
        $this->assertEquals($this->languageArray['scope'], $this->languageObject->getScope());
    }

    /** @test */
    public function it_returns_null_when_missing_scope()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getScope());
    }

    /** @test */
    public function it_returns_type()
    {
        $this->assertEquals($this->languageArray['type'], $this->languageObject->getType());
    }

    /** @test */
    public function it_returns_null_when_missing_type()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getType());
    }

    /** @test */
    public function it_returns_cultures()
    {
        $this->assertEquals($this->languageArray['cultures'], $this->languageObject->getCultures());
    }

    /** @test */
    public function it_returns_null_when_missing_cultures()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getCultures());
    }

    /** @test */
    public function it_returns_culture()
    {
        $this->assertEquals($this->languageArray['cultures']['am-ET'], $this->languageObject->getCulture('am-ET'));
    }

    /** @test */
    public function it_returns_null_when_missing_culture()
    {
        $this->languageObject->setAttributes([]);

        $this->assertNull($this->languageObject->getCulture('am-ET'));
    }
}


tests/Unit/HelpersTest.php
<?php

declare(strict_types=1);

namespace Rinvex\Language\Tests\Unit;

use Rinvex\Language\Language;
use PHPUnit\Framework\TestCase;

class HelpersTest extends TestCase
{
    /** @test */
    public function it_returns_language_data_through_helper()
    {
        $amharic = [
            'name' => 'Amharic',
            'native' => 'አማርኛ',
            'iso_639_1' => 'am',
            'iso_639_2' => 'amh',
            'iso_639_3' => 'amh',
            'script' => [
                'name' => 'Ethiopic (Ge_ez)',
                'iso_15924' => 'Ethi',
                'iso_numeric' => '430',
                'direction' => 'ltr',
            ],
            'family' => [
                'name' => 'Afro-Asiatic',
                'iso_639_5' => 'afa',
                'hierarchy' => 'afa',
            ],
            'cultures' => [
                'am-ET' => [
                    'name' => 'Amharic (Ethiopia)',
                    'native' => 'አማርኛ (ኢትዮጵያ)',
                ],
            ],
            'scope' => 'individual',
            'type' => 'living',
        ];

        $this->assertEquals($amharic, language('am', false));
        $this->assertEquals(new Language($amharic), language('am'));
    }

    /** @test */
    public function it_returns_languages_array_through_helper()
    {
        $this->assertEquals(183, count(languages()));
        $this->assertIsArray(languages()['en']);
        $this->assertEquals('English', languages()['en']['name']);
    }

    /** @test */
    public function it_returns_language_scripts_array_through_helper()
    {
        $this->assertEquals(29, count(language_scripts()));
        $this->assertIsArray(language_scripts());
        $this->assertArrayHasKey('Arab', language_scripts());
    }

    /** @test */
    public function it_returns_language_families_array_through_helper()
    {
        $this->assertEquals(27, count(language_families()));
        $this->assertIsArray(language_families());
        $this->assertArrayHasKey('afa', language_families());
    }
}


src/helpers.php
<?php

declare(strict_types=1);

use Rinvex\Language\LanguageLoader;

if (! function_exists('language')) {
    /**
     * Get the language by it's ISO 639-1.
     *
     * @param string $code
     * @param bool   $hydrate
     *
     * @return \Rinvex\Language\Language|array
     */
    function language($code, $hydrate = true)
    {
        return LanguageLoader::language($code, $hydrate);
    }
}

if (! function_exists('languages')) {
    /**
     * Get all languages.
     *
     * @param bool $hydrate
     *
     * @return array
     */
    function languages($hydrate = false)
    {
        return LanguageLoader::languages($hydrate);
    }
}

if (! function_exists('language_scripts')) {
    /**
     * Get all language scripts.
     *
     * @return array
     */
    function language_scripts()
    {
        return LanguageLoader::scripts();
    }
}

if (! function_exists('language_families')) {
    /**
     * Get all language families.
     *
     * @return array
     */
    function language_families()
    {
        return LanguageLoader::families();
    }
}


src/Language.php
<?php

declare(strict_types=1);

namespace Rinvex\Language;

use Exception;

class Language
{
    /**
     * The attributes array.
     *
     * @var array
     */
    protected $attributes;

    /**
     * Create a new Language instance.
     *
     * @param array $attributes
     *
     * @throws \Exception
     */
    public function __construct($attributes)
    {
        // Set the attributes
        $this->setAttributes($attributes);

        // Check required mandatory attributes
        if (empty($this->getName()) || empty($this->getNativeName()) || empty($this->getIso6391())) {
            throw new Exception('Missing mandatory language attributes!');
        }
    }

    /**
     * Set the attributes.
     *
     * @param array $attributes
     *
     * @return $this
     */
    public function setAttributes($attributes)
    {
        $this->attributes = $attributes;

        return $this;
    }

    /**
     * Get the attributes.
     *
     * @return array|null
     */
    public function getAttributes(): ?array
    {
        return $this->attributes;
    }

    /**
     * Set single attribute.
     *
     * @param string $key
     * @param mixed  $value
     *
     * @return $this
     */
    public function set($key, $value)
    {
        $this->attributes[$key] = $value;

        return $this;
    }

    /**
     * Get an item from attributes array using "dot" notation.
     *
     * @param string $key
     * @param mixed  $default
     *
     * @return mixed
     */
    public function get($key, $default = null)
    {
        $array = $this->attributes;

        if (is_null($key)) {
            return $array;
        }

        if (array_key_exists($key, $array)) {
            return $array[$key];
        }

        foreach (explode('.', $key) as $segment) {
            if (is_array($array) && array_key_exists($segment, $array)) {
                $array = $array[$segment];
            } else {
                return $default;
            }
        }

        return $array;
    }

    /**
     * Get the name.
     *
     * @return string|null
     */
    public function getName(): ?string
    {
        return $this->get('name');
    }

    /**
     * Get the given native name or fallback to first native name.
     *
     * @return string|null
     */
    public function getNativeName(): ?string
    {
        return $this->get('native');
    }

    /**
     * Get the ISO 639-1 code.
     *
     * @return string|null
     */
    public function getIso6391()
    {
        return $this->get('iso_639_1');
    }

    /**
     * Get the ISO 639-2 code.
     *
     * @TODO: add support for both 639-2/T and 639-2/B
     *        https://github.com/rinvex/languages/issues/19
     *
     * @return string|null
     */
    public function getIso6392()
    {
        return $this->get('iso_639_2');
    }

    /**
     * Get the ISO 639-3 code.
     *
     * @return string|null
     */
    public function getIso6393()
    {
        return $this->get('iso_639_3');
    }

    /**
     * Get the script.
     *
     * @return array|null
     */
    public function getScript(): ?array
    {
        return $this->get('script');
    }

    /**
     * Get the script name.
     *
     * @return string|null
     */
    public function getScriptName(): ?string
    {
        return $this->get('script.name');
    }

    /**
     * Get the script ISO 15924.
     *
     * @return string|null
     */
    public function getScriptIso15924()
    {
        return $this->get('script.iso_15924');
    }

    /**
     * Get the script ISO numeric.
     *
     * @return string|null
     */
    public function getScriptIsoNumeric(): ?string
    {
        return $this->get('script.iso_numeric');
    }

    /**
     * Get the script direction.
     *
     * @return string|null
     */
    public function getScriptDirection(): ?string
    {
        return $this->get('script.direction');
    }

    /**
     * Get the family.
     *
     * @return array|null
     */
    public function getFamily(): ?array
    {
        return $this->get('family');
    }

    /**
     * Get the family name.
     *
     * @return string|null
     */
    public function getFamilyName(): ?string
    {
        return $this->get('family.name');
    }

    /**
     * Get the family ISO 639-5.
     *
     * @return string|null
     */
    public function getFamilyIso6395()
    {
        return $this->get('family.iso_639_5');
    }

    /**
     * Get the family hierarchy.
     *
     * @return string|null
     */
    public function getFamilyHierarchy(): ?string
    {
        return $this->get('family.hierarchy');
    }

    /**
     * Get the scope.
     *
     * @return string|null
     */
    public function getScope(): ?string
    {
        return $this->get('scope');
    }

    /**
     * Get the type.
     *
     * @return string|null
     */
    public function getType(): ?string
    {
        return $this->get('type');
    }

    /**
     * Get the cultures.
     *
     * @return array|null
     */
    public function getCultures(): ?array
    {
        return $this->get('cultures');
    }

    /**
     * Get the given culture.
     *
     * @param string|null $culture
     *
     * @return array|null
     */
    public function getCulture($culture = null): ?array
    {
        return $this->getCultures()[$culture] ?? (! empty($this->getCultures()) ? current($this->getCultures()) : null);
    }
}
